ext/winevt/winevt_subscribe.c

```diff
@@ -355,6 +357,19 @@ rb_winevt_subscribe_next(VALUE self)
     return Qfalse;
   }
 
+  /* If a signalEvent notifies a failure, raise
+   * SubscribeHandlerError to detect stale subscription.
+   * Note that we don't need to wait explicitly here.
+   * Because this function is inside of each enumerator.
+   * So, WaitForSingleObject should return immediately and should be
+   * processed with the latter each loop if there is no more items. */
+  dwWait = WaitForSingleObject(winevtSubscribe->signalEvent, 0);
```

Comment:
We should leave a comment that it is possible that it does not actually notify a failure and it is for just in case.

```suggestion
  /* If a signalEvent notifies a failure, raise
   * SubscribeHandlerError to detect stale subscription.
   * Note that it is unclear if the signalEvent actually notifies
   * a failure and this confirming is just in case.
   * It just checks there are no errors at this point, so there is
   * no need to wait for the signal here. */
  dwWait = WaitForSingleObject(winevtSubscribe->signalEvent, 0);
```

As far as I can tell from the documentation, there is no specification that allows the notifying side to cause `WAIT_FAILED` intentionally.
(`Event` is binary information. It just manages `Signaled` or not. It looks `WAIT_FAILED` is normally caused by problems on the confirming side, such as [Access Rights](https://learn.microsoft.com/en-us/windows/win32/sync/synchronization-object-security-and-access-rights).)

I can't deny the possibility that `EvtSubscribe` may notify a failure by `Event` as a special case, but notifying a failure is not mentioned in the documentation at all.
The document only says that it will be `Signaled` when new data becomes available.

I believe the reason for managing `Event` in [the example](https://learn.microsoft.com/en-us/windows/win32/wes/subscribing-to-events#pull-subscriptions) is to wait for new data, not to handle errors.
(Because there is no mention anywhere that `EvtSubscribe` notifies errors with `Event`.)

So, I understand that it's not clear if this code actually works as expected, and it is for just in case.
Based on this understanding, it would be better to have such comments.

Sorry if I have missed something. If so, please point it out to me.

Reply:
I understand your concern. So, we just need to show the raising exception is intended for the further investigation.

Reply:
Thanks!